frontend/src/plugins/impl/panel/PanelPlugin.tsx

```diff
@@ -0,0 +1,215 @@
+interface Data {
+  extension: string | null;
+  docs_json: any;
+  render_json: any;
+}
+
+type T = Record<string, any>;
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type PluginFunctions = {
+  send_to_widget: <T>(req: { message?: any, buffers?: any }) => Promise<null | undefined>;
+};
+
+export const PanelPlugin = createPlugin<T>("marimo-panel")
+  .withData(
+    z.object({
+      extension: z.string().nullable(),
+      docs_json: z.any(),
+      render_json: z.any()
+    }),
+  )
+  .withFunctions<PluginFunctions>({
+    send_to_widget: rpc
+      .input(z.object({ message: z.any(), buffers: z.array(z.any()) }))
+      .output(z.null().optional()),
+  })
+  .renderer((props) => <PanelSlot {...props} />);
+
+
+function isObject(obj: unknown): obj is object {
+  const tp = typeof obj
+  return tp === "function" || tp === "object" && !!obj
+}
+
+function is_nullish(obj: unknown): obj is null | undefined {
+  return obj == null
+}
+
+function isPlainObject<T>(obj: unknown): obj is {[key: string]: T} {
+  return isObject(obj) && (is_nullish(obj.constructor) || obj.constructor === Object)
+}
+
+function extract_buffers(value: unknown, buffers: ArrayBuffer[]): any {
+  if (Array.isArray(value)) {
+    for (const val of value) {
+      extract_buffers(val, buffers)
+    }
+  } else if (value instanceof Map) {
+    for (const key of value.keys()) {
+      const v = value.get(key)
+      extract_buffers(v, buffers)
+    }
+  } else if (value.to_base64 !== undefined) {
+    const {buffer} = value
+    const id = buffers.length
+    buffers.push(buffer)
+    return {id}
+  } else if (isPlainObject(value)) {
+    for (const key of Object.keys(value)) {
+      const replaced = extract_buffers(value[key], buffers)
+      if (replaced != null) {
+        value[key] = replaced
+      }
+    }
+  }
+}
+
+type Props = IPluginProps<T, Data, PluginFunctions>;
+
+const PanelSlot = (props: Props) => {
+  const { extension, docs_json, render_json } = props.data;
+  const ref = useRef<HTMLDivElement>(null);
+  const doc = useRef<any>(null);
+  const receiver = useRef<any>(null);
+  const [loaded, setLoaded] = useState<boolean>(false);
+  const [mounted, setMounted] = useState<boolean>(false);
+  const event_buffer = []
+  let blocked = false
+  let timeout = Date.now()
+
+  useOnMount(() => {
+    if (!ref.current) {
+      return;
+    }
+    setMounted(true)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  });
+
+  useEffect(() => {
+    if (extension.length !== 0) {
+      const script = document.createElement('script');
+      script.innerHTML = extension
+      document.head.appendChild(script);
+    }
+
+    const checkBokeh = setInterval(() => {
+      if (window.Bokeh) {
+        setLoaded(true);
+        clearInterval(checkBokeh);
+      }
+    }, 10);
+
+    return () => {
+      if (extension.length !== 0) {
+	document.head.removeChild(script);
+      }
+      clearInterval(checkBokeh);
+    };
+  }, []);
```

Comment:
i can make some changes after this merged for the outstanding issues. i think we can just register this once, and leave it there for the duration of the user's session (instead of cleaning up)

Reply:
Feel free to push to the PR as well. Intend to spend a little bit more time on this in the morning.